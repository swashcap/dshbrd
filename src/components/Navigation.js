// @flow
import * as React from "react";
import styled from "styled-components";

type Props = {
  initialIndex?: number,
  items: string[],
  onPress: (index: number, event: SyntheticEvent<HTMLAnchorElement>) => void
};

type State = {
  activeIndex: number
};

const NavigationWrapper = styled.nav`
  display: flex;
  justify-content: center;
`;

const NavigationList = styled.ul`
  background: rgba(0, 0, 0, 0.1);
  border-radius: 6px;
  display: flex;
  list-style: none;
  margin: 0;
  padding: 0 1em;
`;

const NavigationListItem = styled.li``;

const NavigationAnchor = styled.a`
  cursor: pointer;
  display: block;
  padding: 1em 0.667em;

  &:focus::before,
  &:hover::before {
    background: ${({ active }) => (active ? "black" : "rgba(0, 0, 0, .1)")};
  }
  &:active::before {
    background: ${({ active }) => (active ? "black" : "rgba(0, 0, 0, .2)")};
  }

  ::before {
    background: ${({ active }) => (active ? "black" : "transparent")};
    border-color: ${({ active }) => (active ? "black" : "rgba(0, 0, 0, .5)")};
    border-radius: 100%;
    border-style: solid;
    border-width: 1px;
    content: "";
    display: inline-block;
    height: 0.5em;
    line-height: 1;
    transition-duration: 200ms;
    duration-property: background border-color;
    vertical-align: middle;
    width: 0.5em;
  }
`;

export default class Navigation extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      activeIndex: props.initialIndex ? props.initialIndex : 0
    };
  }

  handlePress = (index: number, event: SyntheticEvent<HTMLAnchorElement>) => {
    this.setState({ activeIndex: index });
    this.props.onPress(index, event);
  };

  render() {
    const { items, ...rest } = this.props;

    return (
      <NavigationWrapper {...rest}>
        <NavigationList>
          {items.map((item, index) => (
            <NavigationListItem key={item}>
              <NavigationAnchor
                active={index === this.state.activeIndex}
                onClick={event => this.handlePress(index, event)}
              />
            </NavigationListItem>
          ))}
        </NavigationList>
      </NavigationWrapper>
    );
  }
}
