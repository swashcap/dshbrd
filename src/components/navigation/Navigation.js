// @flow
import * as React from "react";

import NavigationWrapper from "./NavigationWrapper";
import NavigationList from "./NavigationList";

type Props = {
  children?: React.Node
};

export default class Navigation extends React.Component<Props> {
  render() {
    const { children, ...rest } = this.props;

    return (
      <NavigationWrapper {...rest}>
        <NavigationList>
          {React.Children.map(children, child => (
            <li key={child.props.to}>{child}</li>
          ))}
        </NavigationList>
      </NavigationWrapper>
    );
  }
}
