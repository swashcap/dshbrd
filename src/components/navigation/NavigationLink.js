// @flow
import * as React from "react";
import { NavLink } from "react-router-dom";
import styled from "styled-components";

import VisuallyHidden from "../VisuallyHidden";

const ACTIVE_CLASS_NAME = "active";

type Props = {
  children?: React.Node,
  to: string
};

/**
 * Make react-router's `NavLink` play well with styled-components.
 */
const _NavigationLink = ({ children, to, ...rest }: Props) => (
  <NavLink activeClassName={ACTIVE_CLASS_NAME} exact to={to} {...rest}>
    <VisuallyHidden>{children}</VisuallyHidden>
  </NavLink>
);

const NavigationLink = styled(_NavigationLink)`
  cursor: pointer;
  display: block;
  padding: 1em 0.667em;

  &:focus::before,
  &:hover::before {
    background: rgba(0, 0, 0, 0.1);
  }
  &:active::before {
    background: rgba(0, 0, 0, 0.2);
  }

  &::before {
    background: transparent;
    border-color: rgba(0, 0, 0, 0.5);
    border-radius: 100%;
    border-style: solid;
    border-width: 1px;
    content: "";
    display: inline-block;
    height: 0.5em;
    line-height: 1;
    transition-duration: 200ms;
    duration-property: background border-color;
    vertical-align: middle;
    width: 0.5em;
  }

  &.${ACTIVE_CLASS_NAME}:focus::before, &.${ACTIVE_CLASS_NAME}:hover::before {
    background: black;
  }
  &.${ACTIVE_CLASS_NAME}:active::before {
    background: black;
  }
  &.${ACTIVE_CLASS_NAME}::before {
    background: black;
    border-color: black;
  }
`;

export default NavigationLink;
