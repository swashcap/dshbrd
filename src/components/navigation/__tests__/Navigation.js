// @flow
import * as React from "react";
import TestRenderer from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";
import { mount } from "enzyme";

import Navigation from "../Navigation";
import NavigationLink from "../NavigationLink";

test("renders correctly", () => {
  const testRenderer = TestRenderer.create(<Navigation />);
  expect(testRenderer).toMatchSnapshot();
});

test("passes items", () => {
  const wrapper = mount(
    <MemoryRouter>
      <Navigation>
        <NavigationLink to="/one">one</NavigationLink>
        <NavigationLink to="/two">two</NavigationLink>
        <NavigationLink to="/three">three</NavigationLink>
      </Navigation>
    </MemoryRouter>
  );

  expect(wrapper.find("a").length).toBe(3);
});
