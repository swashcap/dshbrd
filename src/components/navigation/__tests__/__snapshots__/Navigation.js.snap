// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  background: rgba(0,0,0,0.1);
  border-radius: 6px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  list-style: none;
  margin: 0;
  padding: 0 1em;
}

<nav
  className="c0"
>
  <ul
    className="c1"
  />
</nav>
`;
